src/games/strategy/ui/SwingComponents.java

```diff
@@ -278,4 +290,102 @@ public static JMenu newJMenu(final String menuTitle, final KeyboardCode keyboard
     menu.setMnemonic(keyboardCode.getSwingKeyEventCode());
     return menu;
   }
+
+  /**
+   * Displays a file chooser from which the user can select a file to save.
+   *
+   * <p>
+   * The user will be asked to confirm the save if the selected file already exists.
+   * </p>
+   *
+   * @param parent Determines the {@code Frame} in which the dialog is displayed; if {@code null}, or if {@code parent}
+   *        has no {@code Frame}, a default {@code Frame} is used.
+   * @param fileFilter The file filter initially selected in the file chooser or empty to use no filter; must not be
+   *        {@code null}.
+   * @param defaultExtension The extension to append to the file name if one is not provided by the user or empty to not
+   *        append an extension; must not be {@code null}.
+   *
+   * @return The file selected by the user or empty if the user aborted the save; never {@code null}.
+   */
+  public static Optional<File> promptSaveFile(final Component parent, final Optional<FileFilter> fileFilter,
+      final Optional<String> defaultExtension) {
+    checkNotNull(fileFilter);
+    checkNotNull(defaultExtension);
+
+    final JFileChooser fileChooser = new JFileChooser() {
+      private static final long serialVersionUID = -136588718021703367L;
+
+      @Override
+      public void approveSelection() {
+        defaultExtension.ifPresent(extension -> {
+          setSelectedFile(FileUtil.appendExtensionIfAbsent(getSelectedFile(), extension));
+        });
+
+        final File file = getSelectedFile();
+        if (file.exists()) {
+          final int result = JOptionPane.showConfirmDialog(
+              parent,
+              String.format("A file named \"%s\" already exists. Do you want to replace it?", file.getName()),
+              "Confirm Save",
+              JOptionPane.YES_NO_OPTION,
+              JOptionPane.WARNING_MESSAGE);
+          if (result != JOptionPane.YES_OPTION) {
+            return;
+          }
+        }
+
+        super.approveSelection();
+      }
+    };
+
+    fileFilter.ifPresent(filter -> fileChooser.setFileFilter(filter));
+
+    final int result = fileChooser.showSaveDialog(parent);
+    return (result == JFileChooser.APPROVE_OPTION) ? Optional.of(fileChooser.getSelectedFile()) : Optional.empty();
+  }
+
+  /**
+   * Runs the specified task on a background thread while displaying a progress dialog.
+   *
+   * @param<T> The type of the task result.
+   *
+   * @param frame The {@code Frame} from which the progress dialog is displayed or {@code null} to use a shared, hidden
+   *        frame as the owner of the progress dialog.
+   * @param message The message to display in the progress dialog; must not be {@code null}.
+   * @param task The task to be executed; must not be {@code null}.
+   *
+   * @return A promise that resolves to the result of the task; never {@code null}.
+   */
+  public static <T> CompletableFuture<T> runWithProgressBar(
```

Comment:
Are you referring to the return value of `runWithProgressBar` or the return value of the `Callable` passed as the `task` parameter of `runWithProgressBar`?

If the former, then the answer is yes.  That is how we are able to chain the behavior to be run after the task is complete (i.e. via `whenComplete`).

If the latter, then it's a good point, and the answer is maybe.  Let me explain how it came about...

I originally added an overload for `runWithProgressBar` that accepted a `Runnable` for the case where no return value from the background task is needed (such as the screenshot export code under discussion):

    public static CompletableFuture<Void> runWithProgressBar(
            final Frame frame,
            final String message,
            final Runnable task) {
        return runWithProgressBar(frame, message, () -> {
            task.run();
            return null;
        });
    }

However, I ran into the classic problem of how to handle a checked exception thrown within the `Runnable` (specifically, in this case, because `ScreenshotExporter#runSave` may throw `IOException`).  I considered the following possible solutions to this problem, but none of them seemed satisfactory:

- tunnel the checked exception through an unchecked exception
- introduce a new interface (e.g. `ThrowingRunnable`) that mimics `Runnable` but allows a checked exception to be thrown from `run`
- using a [sneaky throw](http://www.philandstuff.com/2012/04/28/sneakily-throwing-checked-exceptions.html)

In the end, I simply chose to use a `Callable` so that methods that effectively return `void` are represented as `Callable<Void>`, with the ugly `return null` as you observed.

If you prefer one of the alternatives I listed above, or have another suggestion, I'm happy to change it.